perf-regression/gpu-margo-p2p-bw.cu

```diff
@@ -0,0 +1,796 @@
+    for (i = 0; i < g_opts.concurrency; i++) {
+        ABT_thread_join(tid_array[i]);
+        ABT_thread_free(&tid_array[i]);
+
+        out.bytes_moved += arg_array[i].bytes_moved;
+    }
+
+    margo_respond(handle, &out);
+
+    if (in.op == HG_BULK_PULL) {
+        /* calculate how many bytes of the buffer have been transferred */
+        bytes_to_check
+            = (g_opts.g_buffer_size / g_opts.xfer_size) * g_opts.xfer_size;
+        if (out.bytes_moved < bytes_to_check) bytes_to_check = out.bytes_moved;
+
+        /* check integrity of fill pattern.  Note that this isn't as strong as
+         * checking every RDMA transfer separately since we are looping around
+         * and overwriting in a ring-buffer style.  We could corrupt early and
+         * but then overwrite it with correct results on a later pass.
+         */
+	/* Perform cudaMemcpy only if (GPU to GPU  || -k option) */
+	if (use_cuda_buf)
+    	    cudaMemcpy(g_buffer, c_buffer, g_opts.g_buffer_size, cudaMemcpyDeviceToHost);
+
+        for (x = 0; x < (bytes_to_check / sizeof(x)); x++) {
+            assert(((hg_size_t*)g_buffer)[x] == x);
+        }
+
+        /* fill pattern for return trip, increment each value by 1 */
+        for (x = 0; x < (g_opts.g_buffer_size / sizeof(x)); x++)
+            ((hg_size_t*)g_buffer)[x] = x + 1;
+
+	/* Perform cudaMemcpy only if (GPU to GPU || -j option) */
+	if(use_cuda_buf)
+    	    cudaMemcpy(c_buffer, g_buffer, g_opts.g_buffer_size, cudaMemcpyHostToDevice);
+    }
+
+    margo_free_input(handle, &in);
+    margo_destroy(handle);
+
+    free(tid_array);
+
+    ABT_mutex_free(&cur_off_mutex);
+
+    if (in.shutdown) ABT_eventual_set(g_bw_done_eventual, NULL, 0);
+
+    free(arg_array);
+
+    return;
+}
+DEFINE_MARGO_RPC_HANDLER(bw_ult)
+
+static int run_benchmark(hg_id_t           id,
+                         ssg_member_id_t   target,
+                         ssg_group_id_t    gid,
+                         margo_instance_id mid,
+                         int               shutdown,
+                         int               duration,
+                         int               print_flag)
+{
+    hg_handle_t  handle;
+    hg_addr_t    target_addr;
+    int          ret;
+    bw_rpc_in_t  in;
+    bw_rpc_out_t out;
+    void*        buffer = g_buffer;
+    hg_size_t    i;
+    hg_size_t    bytes_to_check;
+    double       start_ts, end_ts;
+    struct hg_bulk_attr bulk_attr;
+
+    /* fill pattern in origin buffer */
+    for (i = 0; i < (g_opts.g_buffer_size / sizeof(i)); i++)
+        ((hg_size_t*)buffer)[i] = i;
+
+    /* Perform cudaMemcpy only if (GPU to GPU || -k option) */
+    if (use_cuda_buf) 
+    	cudaMemcpy(c_buffer, buffer, g_opts.g_buffer_size, cudaMemcpyHostToDevice);
+
+    ret = ssg_get_group_member_addr(gid, target, &target_addr);
+    assert(ret == SSG_SUCCESS);
+
+    ret = margo_create(mid, target_addr, id, &handle);
+    assert(ret == 0);
+
+    bulk_attr.mem_type = (hg_mem_type_t)NA_MEM_TYPE_CUDA;
```

Comment:
Should we only be setting the mem_type as NA_MEM_TYPE_CUDA in the case where we're actually using a GPU buffer?

Reply:
Fixed